diff --git a/gaps/gpu-shower/main.cu b/gaps/gpu-shower/main.cu
index 14b3a9b..d4e36e9 100644
--- a/gaps/gpu-shower/main.cu
+++ b/gaps/gpu-shower/main.cu
@@ -90,8 +90,7 @@ void run_generator(const params& p) {
   auto start = std::chrono::high_resolution_clock::now();
 
   // Calculate the leading order cross section and kinematics
-  calc_lome(dv_events, p.process, p.nlo, p.root_s, p.asmz, blocks, p.threads,
-            p.me2pdf);
+  calc_lome(dv_events, p, blocks);
 
   auto end = std::chrono::high_resolution_clock::now();
   std::chrono::duration<double> diff_me = end - start;
@@ -107,9 +106,7 @@ void run_generator(const params& p) {
     std::cout << "Showering partons..." << std::endl;
     start = std::chrono::high_resolution_clock::now();
 
-    run_shower(dv_events, p.root_s, p.nlo, p.do_partitioning, p.t_c, p.asmz,
-               p.fixed_as, p.use_cmw, p.n_emissions_max, blocks, p.threads,
-               p.showerpdf);
+    run_shower(dv_events, p, blocks);
 
     end = std::chrono::high_resolution_clock::now();
     diff_sh = end - start;
@@ -124,7 +121,7 @@ void run_generator(const params& p) {
   start = std::chrono::high_resolution_clock::now();
 
   // analysis
-  do_analysis(dv_events, p.storage_file, p.process, blocks, p.threads);
+  do_analysis(dv_events, p, blocks);
 
   end = std::chrono::high_resolution_clock::now();
   std::chrono::duration<double> diff_an = end - start;
diff --git a/gaps/gpu-shower/matrix/include/matrix.cuh b/gaps/gpu-shower/matrix/include/matrix.cuh
index 946aa6f..ffd2608 100644
--- a/gaps/gpu-shower/matrix/include/matrix.cuh
+++ b/gaps/gpu-shower/matrix/include/matrix.cuh
@@ -2,6 +2,7 @@
 #define matrix_cuh
 
 #include "event.cuh"
+#include "interface.cuh"
 #include "pdf.cuh"
 #include "qcd.cuh"
 
@@ -71,8 +72,7 @@ void lhc_nlo(thrust::device_vector<event>& d_events, matrix* matrix,
              alpha_s* as, pdf_wrapper* pdf, int blocks, int threads);
 
 // all tasks wrapped in a function
-void calc_lome(thrust::device_vector<event>& d_events, int process, bool nlo,
-               double root_s, double asmz, int blocks, int threads,
-               const std::string& pdf_name = "CT14lo");
+void calc_lome(thrust::device_vector<event>& d_events, const params& p,
+               int blocks);
 
 #endif  // matrix_cuh
\ No newline at end of file
diff --git a/gaps/gpu-shower/matrix/src/matrix.cu b/gaps/gpu-shower/matrix/src/matrix.cu
index 295e44d..31b3848 100644
--- a/gaps/gpu-shower/matrix/src/matrix.cu
+++ b/gaps/gpu-shower/matrix/src/matrix.cu
@@ -20,9 +20,8 @@ __global__ void matrix_setup_kernel(matrix* matrix, int process, bool nlo,
 // main
 
 // function to generate the lo matrix elements + momenta
-void calc_lome(thrust::device_vector<event>& d_events, int process, bool nlo,
-               double root_s, double asmz, int blocks, int threads,
-               const std::string& pdf_name) {
+void calc_lome(thrust::device_vector<event>& d_events, const params& p,
+               int blocks) {
   /**
    * @brief wrapper for matrix element calculation
    *
@@ -45,37 +44,37 @@ void calc_lome(thrust::device_vector<event>& d_events, int process, bool nlo,
 
   // set up the device matrix object
   debug_msg("running @matrix_setup_kernel");
-  matrix_setup_kernel<<<1, 1>>>(d_matrix, process, nlo, root_s);
+  matrix_setup_kernel<<<1, 1>>>(d_matrix, p.process, p.nlo, p.root_s);
   sync_gpu_and_check("matrix_setup_kernel");
 
   // set up the device alpha_s calculator
   alpha_s* d_as;
   cudaMalloc(&d_as, sizeof(alpha_s));
-  as_setup_kernel<<<1, 1>>>(d_as, asmz);
+  as_setup_kernel<<<1, 1>>>(d_as, p.asmz);
   sync_gpu_and_check("as_setup_kernel");
 
   // set up the pdf evaluator (for LHC processes)
-  pdf_wrapper pdf(pdf_name);
+  pdf_wrapper pdf(p.me2pdf);
 
   // LEP LO
-  if ((process == 1) && !nlo) {
-    lep_lo(d_events, d_matrix, blocks, threads);
+  if ((p.process == 1) && !p.nlo) {
+    lep_lo(d_events, d_matrix, blocks, p.threads);
   }
 
   // LEP NLO
-  else if ((process == 1) && nlo) {
-    lep_lo(d_events, d_matrix, blocks, threads);
-    lep_nlo(d_events, d_matrix, d_as, blocks, threads);
+  else if ((p.process == 1) && p.nlo) {
+    lep_lo(d_events, d_matrix, blocks, p.threads);
+    lep_nlo(d_events, d_matrix, d_as, blocks, p.threads);
   }
 
   // LHC LO - and just do NLO = LO for now
-  else if ((process == 2) && !nlo) {
-    lhc_lo(d_events, d_matrix, &pdf, blocks, threads);
+  else if ((p.process == 2) && !p.nlo) {
+    lhc_lo(d_events, d_matrix, &pdf, blocks, p.threads);
   }
 
-  else if ((process == 2) && nlo) {
-    lhc_lo(d_events, d_matrix, &pdf, blocks, threads);
-    lhc_nlo(d_events, d_matrix, d_as, &pdf, blocks, threads);
+  else if ((p.process == 2) && p.nlo) {
+    lhc_lo(d_events, d_matrix, &pdf, blocks, p.threads);
+    lhc_nlo(d_events, d_matrix, d_as, &pdf, blocks, p.threads);
   }
 
   return;
diff --git a/gaps/gpu-shower/observables/include/observables.cuh b/gaps/gpu-shower/observables/include/observables.cuh
index fb148ea..930bc5f 100644
--- a/gaps/gpu-shower/observables/include/observables.cuh
+++ b/gaps/gpu-shower/observables/include/observables.cuh
@@ -3,6 +3,7 @@
 
 #include "event.cuh"
 #include "histogram.cuh"
+#include "interface.cuh"
 
 // add analyses here
 #include "eventshapes.cuh"
@@ -86,7 +87,7 @@ __global__ void fill_histos(analysis* an, const event* events, double* results,
                             int process, int n);
 
 // analysis wrapped in a function
-void do_analysis(thrust::device_vector<event>& d_events, std::string filename,
-                 int process, int blocks, int threads);
+void do_analysis(thrust::device_vector<event>& d_events, const params& p,
+                 int blocks);
 
 #endif  // observables_cuh_
\ No newline at end of file
diff --git a/gaps/gpu-shower/observables/src/observables.cu b/gaps/gpu-shower/observables/src/observables.cu
index ab6b830..d5cf8ea 100644
--- a/gaps/gpu-shower/observables/src/observables.cu
+++ b/gaps/gpu-shower/observables/src/observables.cu
@@ -142,8 +142,8 @@ void write_xsec(double xsec, double xsec_err, const std::string& filename) {
 
 // -----------------------------------------------------------------------------
 // run the above kernels
-void do_analysis(thrust::device_vector<event>& dv_events, std::string filename,
-                 int process, int blocks, int threads) {
+void do_analysis(thrust::device_vector<event>& dv_events, const params& p,
+                 int blocks) {
   /**
    * @brief Run the analysis
    *
@@ -156,7 +156,7 @@ void do_analysis(thrust::device_vector<event>& dv_events, std::string filename,
   analysis *h_an, *d_an;
 
   // allocate memory for the device analysis object
-  h_an = new analysis(process);
+  h_an = new analysis(p.process);
   cudaMalloc(&d_an, sizeof(analysis));
   cudaMemcpy(d_an, h_an, sizeof(analysis), cudaMemcpyHostToDevice);
 
@@ -169,7 +169,7 @@ void do_analysis(thrust::device_vector<event>& dv_events, std::string filename,
   cudaMalloc(&d_invalid, sizeof(int));
   cudaMemset(d_invalid, 0, sizeof(int));
 
-  validate_events<<<blocks, threads>>>(d_events, d_invalid, n_events);
+  validate_events<<<blocks, p.threads>>>(d_events, d_invalid, n_events);
   sync_gpu_and_check("validate_events");
 
   int h_invalid;
@@ -190,32 +190,32 @@ void do_analysis(thrust::device_vector<event>& dv_events, std::string filename,
   thrust::device_vector<double> dv_results(20 * n_events, -50.);
   double* d_results = thrust::raw_pointer_cast(dv_results.data());
 
-  if (process == 1) {
-    cluster_durham<<<blocks, threads>>>(d_events, d_results, n_events);
+  if (p.process == 1) {
+    cluster_durham<<<blocks, p.threads>>>(d_events, d_results, n_events);
     sync_gpu_and_check("cluster_durham");
 
-    calculate_ev_shapes<<<blocks, threads>>>(d_events, d_results, n_events);
+    calculate_ev_shapes<<<blocks, p.threads>>>(d_events, d_results, n_events);
     sync_gpu_and_check("calculate_ev_shapes");
   }
 
-  else if (process == 2) {
-    calculate_mczinc<<<blocks, threads>>>(d_events, d_results, n_events);
+  else if (p.process == 2) {
+    calculate_mczinc<<<blocks, p.threads>>>(d_events, d_results, n_events);
     sync_gpu_and_check("calculate_mczinc");
 
-    cluster_genkt<<<blocks, threads>>>(d_events, d_results, n_events);
+    cluster_genkt<<<blocks, p.threads>>>(d_events, d_results, n_events);
     sync_gpu_and_check("cluster_genkt");
   }
 
   // do the analysis
-  fill_histos<<<blocks, threads>>>(d_an, d_events, d_results, process,
-                                   n_events);
+  fill_histos<<<blocks, p.threads>>>(d_an, d_events, d_results, p.process,
+                                     n_events);
   sync_gpu_and_check("fill_histos");
 
   // copy the results back to the host
   cudaMemcpy(h_an, d_an, sizeof(analysis), cudaMemcpyDeviceToHost);
 
   // remove existing file
-  std::remove(filename.c_str());
+  std::remove(p.storage_file.c_str());
 
   // Calculate cross-section
   double xsec = h_an->wtot / h_an->ntot;      // Cross-section in pb
@@ -225,12 +225,12 @@ void do_analysis(thrust::device_vector<event>& dv_events, std::string filename,
   printf("Total cross-section: %.2e nb\n", xsec / 1000.);
 
   // Write cross-section in YODA format
-  write_xsec(xsec, xsec_err, filename);
+  write_xsec(xsec, xsec_err, p.storage_file);
 
   for (auto& hist : h_an->hists) {
     if (hist.name[0] != 'h') {
       hist.scale_w(1. / h_an->wtot);
-      write(hist, filename);
+      write(hist, p.storage_file);
     }
   }
 
diff --git a/gaps/gpu-shower/shower/include/shower.cuh b/gaps/gpu-shower/shower/include/shower.cuh
index ab835b4..2db5be7 100644
--- a/gaps/gpu-shower/shower/include/shower.cuh
+++ b/gaps/gpu-shower/shower/include/shower.cuh
@@ -1,6 +1,7 @@
 #ifndef shower_cuh_
 #define shower_cuh_
 
+#include "interface.cuh"
 #include "pdf.cuh"
 #include "qcd.cuh"
 
@@ -123,10 +124,7 @@ __global__ void check_too_many_particles(event* events, int n_emissions_max,
                                          int* d_completed, int n);
 
 // all tasks wrapped into a function
-void run_shower(thrust::device_vector<event>& dv_events, double root_s,
-                bool nlo_matching, bool do_partitioning, double t_c,
-                double asmz, bool fixed_as, bool use_cmw,
-                int n_emissions_max, int blocks, int threads,
-                const std::string& pdf_name = "CT14lo");
+void run_shower(thrust::device_vector<event>& dv_events, const params& p,
+                int blocks);
 
 #endif  // shower_cuh_
\ No newline at end of file
diff --git a/gaps/gpu-shower/shower/src/shower.cu b/gaps/gpu-shower/shower/src/shower.cu
index 817c3b1..38f9ef7 100644
--- a/gaps/gpu-shower/shower/src/shower.cu
+++ b/gaps/gpu-shower/shower/src/shower.cu
@@ -602,11 +602,8 @@ struct is_not_end_shower {
 
 // -----------------------------------------------------------------------------
 
-void run_shower(thrust::device_vector<event>& dv_events, double root_s,
-                bool nlo_matching, bool do_partitioning, double t_c,
-                double asmz, bool fixed_as, bool use_cmw,
-                int n_emissions_max, int blocks, int threads,
-                const std::string& pdf_name) {
+void run_shower(thrust::device_vector<event>& dv_events, const params& p,
+                int blocks) {
   /**
    * @brief Run the shower on the events
    *
@@ -623,13 +620,13 @@ void run_shower(thrust::device_vector<event>& dv_events, double root_s,
   // set up the device alpha_s calculator
   alpha_s* d_as;
   cudaMalloc(&d_as, sizeof(alpha_s));
-  as_setup_kernel<<<1, 1>>>(d_as, asmz, (fixed_as ? 0 : 2), use_cmw);
+  as_setup_kernel<<<1, 1>>>(d_as, p.asmz, (p.fixed_as ? 0 : 2), p.use_cmw);
   sync_gpu_and_check("as_setup_kernel");
 
   // Calculate as_max = as(t_c)
   double* d_as_max;
   cudaMalloc(&d_as_max, sizeof(double));
-  as_value<<<1, 1>>>(d_as, d_as_max, t_c);
+  as_value<<<1, 1>>>(d_as, d_as_max, p.t_c);
   sync_gpu_and_check("as_value");
   double as_max;
   cudaMemcpy(&as_max, d_as_max, sizeof(double), cudaMemcpyDeviceToHost);
@@ -637,11 +634,11 @@ void run_shower(thrust::device_vector<event>& dv_events, double root_s,
   // set up the shower
   shower* d_shower;
   cudaMalloc(&d_shower, sizeof(shower));
-  shower_setup_kernel<<<1, 1>>>(d_shower, root_s, t_c, as_max);
+  shower_setup_kernel<<<1, 1>>>(d_shower, p.root_s, p.t_c, as_max);
   sync_gpu_and_check("shower_setup_kernel");
 
   // set up the pdf evaluator
-  pdf_wrapper pdf(pdf_name);
+  pdf_wrapper pdf(p.showerpdf);
 
   /**
    * Shower Variables - useful to store as collective
@@ -696,7 +693,7 @@ void run_shower(thrust::device_vector<event>& dv_events, double root_s,
   // prepare the shower
 
   debug_msg("running @prep_shower");
-  prep_shower<<<blocks, threads>>>(d_events, nlo_matching, n_events);
+  prep_shower<<<blocks, p.threads>>>(d_events, p.nlo, n_events);
   sync_gpu_and_check("prep_shower");
 
   // ---------------------------------------------------------------------------
@@ -708,7 +705,7 @@ void run_shower(thrust::device_vector<event>& dv_events, double root_s,
 
   // (Varying) kernel size and partition factor
   int n = n_events;
-  int p = 1;
+  int part_step = 1;
 
   while (completed < n_events) {
     // run all the kernels here...
@@ -717,51 +714,51 @@ void run_shower(thrust::device_vector<event>& dv_events, double root_s,
     // check if there are too many particles (do first in case of H event)
 
     debug_msg("running @check_too_many_particles");
-    check_too_many_particles<<<blocks, threads>>>(
-        d_events, n_emissions_max, d_too_many_particles, d_completed, n);
+    check_too_many_particles<<<blocks, p.threads>>>(
+        d_events, p.n_emissions_max, d_too_many_particles, d_completed, n);
     sync_gpu_and_check("check_too_many_particles");
 
     // -------------------------------------------------------------------------
     // select the winner kernel
 
     debug_msg("running @select_winner_split_func");
-    select_winner_split_func<<<blocks, threads>>>(d_shower, d_events, n,
-                                                  d_winner);
+    select_winner_split_func<<<blocks, p.threads>>>(d_shower, d_events, n,
+                                                   d_winner);
     sync_gpu_and_check("select_winner_split_func");
 
     // -------------------------------------------------------------------------
     // check cutoff
 
     debug_msg("running @check_cutoff");
-    check_cutoff<<<blocks, threads>>>(d_events, d_shower, d_completed, n);
+    check_cutoff<<<blocks, p.threads>>>(d_events, d_shower, d_completed, n);
     sync_gpu_and_check("check_cutoff");
 
     // -------------------------------------------------------------------------
     // calculate pdf of ij and i using LHAPDF
 
     debug_msg("running @setup_pdfratio");
-    setup_pdfratio<<<blocks, threads>>>(d_shower, d_events, n, d_flavours_a,
-                                        d_flavours_b, d_x_a, d_x_b, d_q2,
-                                        d_winner);
+    setup_pdfratio<<<blocks, p.threads>>>(d_shower, d_events, n, d_flavours_a,
+                                          d_flavours_b, d_x_a, d_x_b, d_q2,
+                                          d_winner);
     sync_gpu_and_check("setup_pdfratio");
 
-    pdf.evaluate(d_flavours_a, d_x_a, d_q2, d_xf_a, n, blocks, threads);
-    pdf.evaluate(d_flavours_b, d_x_b, d_q2, d_xf_b, n, blocks, threads);
+    pdf.evaluate(d_flavours_a, d_x_a, d_q2, d_xf_a, n, blocks, p.threads);
+    pdf.evaluate(d_flavours_b, d_x_b, d_q2, d_xf_b, n, blocks, p.threads);
 
     // -------------------------------------------------------------------------
     // veto algorithm
 
     debug_msg("running @veto_alg");
-    veto_alg<<<blocks, threads>>>(d_shower, d_as, d_events, n, d_xf_a, d_xf_b,
-                                  d_accept_emission, d_winner);
+    veto_alg<<<blocks, p.threads>>>(d_shower, d_as, d_events, n, d_xf_a,
+                                   d_xf_b, d_accept_emission, d_winner);
     sync_gpu_and_check("veto_alg");
 
     // -------------------------------------------------------------------------
     // splitting algorithm
 
     debug_msg("running @do_splitting");
-    do_splitting<<<blocks, threads>>>(d_shower, d_events, n, d_accept_emission,
-                                      d_winner);
+    do_splitting<<<blocks, p.threads>>>(d_shower, d_events, n, d_accept_emission,
+                                        d_winner);
     sync_gpu_and_check("do_splitting");
 
     // -------------------------------------------------------------------------
@@ -773,11 +770,11 @@ void run_shower(thrust::device_vector<event>& dv_events, double root_s,
     // -------------------------------------------------------------------------
 
     // Partition the events based on completion at 50%, 75%, 87.5%, etc.
-    if (do_partitioning &&
-        completed >= static_cast<int>(n_events * (1 - 1 / pow(2, p)))) {
+    if (p.do_partitioning &&
+        completed >= static_cast<int>(n_events * (1 - 1 / pow(2, part_step)))) {
       // Stop at 25k, below this, the overhead of partitioning outweighs the
       // benefits, increasing execution time.
-      if (static_cast<int>(n_events * (1 / pow(2, p))) >= 25000) {
+      if (static_cast<int>(n_events * (1 / pow(2, part_step))) >= 25000) {
         std::cerr << std::endl;
         std::cerr << "partition at " << completed << "/" << n_events
                   << std::endl;
@@ -793,7 +790,7 @@ void run_shower(thrust::device_vector<event>& dv_events, double root_s,
         n = n_events - completed;
 
         // Increment the partitioning step
-        p++;
+        part_step++;
       }
     }
 
@@ -812,7 +809,7 @@ void run_shower(thrust::device_vector<event>& dv_events, double root_s,
   cudaMemcpy(&too_many_particles, d_too_many_particles, sizeof(int),
              cudaMemcpyDeviceToHost);
   if (too_many_particles > 0) {
-    if (max_particles < n_emissions_max) {
+    if (max_particles < p.n_emissions_max) {
       std::cerr << "Warning: " << too_many_particles
                 << " events surpassed the maximum number of particles"
                 << std::endl;
